### ISSUE TITLE
after-session-close not triggering

### ISSUE THREAD

**Issue by farnoy** [tmux/tmux#1245]
Hi,

I'm trying to use `after-session-closed` hook to do something useful, but I can't get it to trigger.

Steps I'm doing:

1. `tmux -v`
1. `<C-b>:set-hook after-session-closed "run-shell \"touch test-file\""<Enter>`
1. `<C-b>:show-hooks<Enter>`
1. observe output: `after-session-closed -> run-shell "touch test-file"`
1. `q`
1. `<C-d>`
1. out of tmux here
1.
    ```
    $ stat test-file
    stat: cannot stat 'test-file': No such file or directory
    ```
Not sure what happens here, I tried `set-hook -t 0`, `-g`, replacing `after-session-closed` with `session-closed`, adding `-b` to `run-shell`. I can't figure out what's wrong.

Formalities:

```sh
$ uname -sp && tmux -V && echo $TERM
Linux unknown
tmux 2.6
xterm-termite
```

Reproduced on git master.

TERM inside is `screen`

I don't have configs, logs are uploaded below.
[tmux-client-84526.log](https://github.com/tmux/tmux/files/1668107/tmux-client-84526.log)
[tmux-server-84528.log](https://github.com/tmux/tmux/files/1668108/tmux-server-84528.log)

(log says loading tmux.config twice, these files do not exist)

**Comment by ThomasAdam** [comment]
There is no `after-session-closed` hook.

I suspect as well that per-session hooks for `session-closed` are broken, given how we handle notifys for session_destroy().  Will look.

For now, you're wanting: `set-hook -g session-closed "run-shell ....."`

**Comment by farnoy** [comment]
@ThomasAdam Your workaround is good only when the current session that is exiting is not the last one on the server. It seems that this hook is not triggered (even in -g mode) if there are no others left.

**Comment by farnoy** [comment]
My current workaround is this (I am trying to come up with a command that would put me in a tmpdir and clean it up when I'm done):

`alias in_tempdir="TMPDIR_THIS=\$(mktemp -d); print \$TMPDIR_THIS; tmux -S \$TMPDIR_THIS/.tmux-sock new-session \; new-session -t in_temp \; send-keys \"pushd \$TMPDIR_THIS\" \; send-keys Enter \; set-hook -gt in_temp session-closed \"run-shell \\\"rm -rf \$TMPDIR_THIS\\\" ; kill-server\" "`

Cleans up the dir and the temporary server, but the downside is no sharing with the default server.

By the way, it seems `set-hook` overrides the previous hook (if it was bound), is that a conscious design decision?

**Comment by ThomasAdam** [comment]
@farnoy Yes, you only get one type of hook.  If you want multiple commands, use them:  `foo ; bar ; baz` -- tmux will interpret them the same, and this design is deliberate.  For now, it mimics how the `set` command works, for example.

As for your observation about `session-closed` hook not working, can you give a simplified version?  I have the `session-closed` hook working for a single session and for global hooks, hence:

```
set-hook -g session-closed 'run-shell "touch /tmp/foo1"'
set-hook session-closed 'run-shell "touch /tmp/foo2'"
```
Both work in the way they ought to.

**Comment by farnoy** [comment]
@ThomasAdam How about this?
```
$ tmux start \; new-session \; set-hook -g session-closed 'run-shell "touch /tmp/test"' \; run-shell "sleep 1" \; send-keys C-d
[exited]
$ stat /tmp/test     
stat: cannot stat '/tmp/test': No such file or directory
```

I had to add the sleep so it sends the keys after shell is initialized. Does this work for you?

**Comment by nicm** [comment]
Try the patch in https://github.com/tmux/tmux/issues/1174 please.

**Comment by farnoy** [comment]
@nicm The patch works for hooks set up with `set-hook -g` (my original example), but it still doesn't work for hooks directly attached to a session:

```
$ ./tmux start \; new-session -s session-one \; set-hook -t session-one session-closed 'run-shell "touch /tmp/test"' \; run-shell "sleep 1" \; send-keys C-d
$ stat /tmp/test     
stat: cannot stat '/tmp/test': No such file or directory
```

**Comment by nicm** [comment]
I am going to apply this so global hooks work and I will take a look at per-session hooks later on.

**Comment by nicm** [comment]
Try this please: [x.diff.txt](https://github.com/tmux/tmux/files/1792302/x.diff.txt)

You'll need to wait for the previous change to make it to GitHub.
